[FIX] Preserve attribute list for MEM_REF expressions of unaligned types

petrokarashchenko opened the pull request:
[FIX] Propagate uncached type attributes to unaligned types

[ARC] Update tests

gcc/
xxxx-xx-xx  Petro Karashchenko  <petro.karashchenko@ringteam.com>

	* emit-rtl.c (adjust_address_1): Preserve attribute list
        for MEM_REF expressions.

testsuite/
xxxx-xx-xx  Petro Karashchenko  <petro.karashchenko@ringteam.com>

        * gcc.target/arc/uncached-9.c: New file.

petrokarashchenko committed:
[FIX] Preserve attribute list for MEM_REF expressions of unaligned types

[FIX] Propagate uncached type attributes to unaligned types

[ARC] Update tests

gcc/
xxxx-xx-xx  Petro Karashchenko  <petro.karashchenko@ringteam.com>

	* emit-rtl.c (adjust_address_1): Preserve attribute list
        for MEM_REF expressions.

testsuite/
xxxx-xx-xx  Petro Karashchenko  <petro.karashchenko@ringteam.com>

        * gcc.target/arc/uncached-9.c: New file.

claziss commented:
This is a mainline modification. Unfortunately, I cannot use it in the official releases. However, you can use it as you please. Thank you for your understanding.

petrokarashchenko commented:
Got it. Thanks!

claziss commented:
Probably your patch can be part of a solution, but, most probably, we can resolve your problem using a movmisalingn pattern directly in backend. But it can lead to some complications :)

petrokarashchenko commented:
I fully understand that I'm a dummy in GCC :) I just got some findings and tried few experiments including `copy_node (MEM_REF)` + `TREE_TYPE (attrs.expr) = TREE_TYPE (TREE_CODE (MEM_EXPR (memref)))`, etc. and variant in this patch seemed to be most safe to me.
Anyway looking forward to get expert fix from you!

claziss commented:
Don't misunderstand me, it is good to try out. Your patch is solving the problem which you observed. You can try out the misalign pattern ;) I can help you as you go.

petrokarashchenko commented:
@claziss I have tried to apply suggestion and dig deeper into misalign pattern direction. But currently it does not seems to be working. I have modified `arc.md` file and added next patterns:
```
(define_expand "movmisalignqi"
 [(set (match_operand:QI 0 "move_dest_operand" "")
       (match_operand:QI 1 "general_operand" ""))]
 ""
 "if (prepare_misaligned_move_operands (operands, QImode)) DONE;")

(define_expand "movmisalignhi"
 [(set (match_operand:HI 0 "move_dest_operand" "")
       (match_operand:HI 1 "general_operand" ""))]
 ""
 "if (prepare_misaligned_move_operands (operands, HImode)) DONE;")

(define_expand "movmisalignsi"
 [(set (match_operand:SI 0 "move_dest_operand" "")
       (match_operand:SI 1 "general_operand" ""))]
 ""
 "if (prepare_misaligned_move_operands (operands, SImode)) DONE;")

(define_expand "movmisaligndi"
 [(set (match_operand:DI 0 "move_dest_operand" "")
       (match_operand:DI 1 "general_operand" ""))]
 ""
 "if (prepare_misaligned_move_operands (operands, DImode)) DONE;")

(define_expand "movmisalignsf"
 [(set (match_operand:SF 0 "move_dest_operand" "")
       (match_operand:SF 1 "general_operand" ""))]
 ""
 "if (prepare_misaligned_move_operands (operands, SFmode)) DONE;")

(define_expand "movmisaligndf"
 [(set (match_operand:DF 0 "move_dest_operand" "")
       (match_operand:DF 1 "general_operand" ""))]
 ""
 "if (prepare_misaligned_move_operands (operands, DFmode)) DONE;")
```
Also I have added prototype of `prepare_misaligned_move_operands()` and next implementation:
```
bool
prepare_misaligned_move_operands (rtx *operands, machine_mode mode)
{
  if (MEM_P (operands[0]))
    {
      fprintf(stderr, "\nUnaligned operands[0]:\n");
      print_rtl_single(stderr, operands[0]);
    }
  if (MEM_P (operands[1]))
    {
      fprintf(stderr, "\nUnaligned operands[1]:\n");
      print_rtl_single(stderr, operands[1]);
    }
  return false;
}
```
But I observe none of my ecpected outputs in `stderr` when I run compiler with test code. Am I doing something in wrong way? Maybe we can continue discussion with an e-mail thread? Or maybe there is a forum so I can create a topic and track discussion?

petrokarashchenko commented:
From GCC internals I read next about movmisaling _m_:
```
This pattern is used by the autovectorizer, and when expanding a MISALIGNED_INDIRECT_REF expression.
```
So we are not having here `autovectorizer` running, and I missed to find in code what is `MISALIGNED_INDIRECT_REF `
